Fix Issue 20074 - header file generation doesn't include attributes with CallExp

RazvanN7 opened the pull request:
I don't know why the constructor body is being outputted. Can anyone shed some light on this?

RazvanN7 committed:
Fix Issue 20074 - header file generation doesn't include attributes with CallExp

rainers commented on src/dmd/hdrgen.d:
```diff
@@ -1606,6 +1606,16 @@ public:
+            buf.writeByte(' ');
+            buf.writestring(str);
+        }
+        tf.attributesApply(&printAttribute);
```
The comment three lines above the patch says that omitting the attributes is deliberate. It actually can be a bit annoying in the error message if the problem is not related to inferred attributes.

RazvanN7 replied:
I agree, but on the other hand it may lead to (falsely) uncompilable code. Consider:

```d
class C
{
    this() @safe
    {
        () @trusted
        {
            int a; int*p = &a;                                                                                                                
        }();
    }   
}
```

Without this patch, the compiler will omit `@trusted` and generate uncompilable code from perfectly fine code.

The comment baffled me also, but I think that this is the correct way to go.

rainers replied:
You should adjust or remove the comment in that case. A compromise might be to emit explicitly specified attributes, but not inferred ones. But the required info might not exist and not worth the complications to add.

rainers commented:
> I don't know why the constructor body is being outputted. Can anyone shed some light on this?

That's because a ctor has no return type and is falsely considered `auto` here;
https://github.com/dlang/dmd/blob/84bcc67667ce944445ac4877e59199c4a10325f4/src/dmd/hdrgen.d#L1487

RazvanN7 commented:
>That's because a ctor has no return type and is falsely considered auto here;

So this is also a bug.

rainers commented:
> So this is also a bug.

Yes, and if it is fixed the test case in this PR might get removed. So I suggest adding a test where the function literal is used by some declaration, e.g. as a template argument.

RazvanN7 commented:
@rainers I changed the test case so that the literal is inside a templated function's body and also I deleted the comment. I think this is good to go now.

rainers reviewed (approved):
Thanks

RazvanN7 commented:
Maybe it's time we get this in?

ZombineDev commented:
Thanks for the reminder ;)
